<?xml version="1.0"?>

<robot name="planar_robot" xmlns:xacro="http://ros.org/wiki/xacro">

    <!-- Constants for robot dimensions -->
    <xacro:property name="radius" value="0.02" />
    <xacro:property name="length" value="0.25" />
    <xacro:property name="damping" value="0.1" /> 
    <xacro:property name="mass" value="0.5" />
    <xacro:property name="effort_limit" value="10" />
    <xacro:property name="velocity_limit" value="100" />

    <xacro:include filename="$(find planar_robot_tutorial)/urdf/materials.xacro" />
    <xacro:include filename="$(find planar_robot_tutorial)/urdf/planar_robot.gazebo" />

    <!-- World -->
    <link name='world' />

    <joint name="joint_base" type="fixed">
        <parent link="world" />
        <child link="base_link"/>
    </joint>

    <!-- Base link -->
    <link name='base_link'>
        <inertial>
            <origin xyz="-0.00 0.0 0.3" rpy="0 0 0" />
            <inertia ixx="1.11" ixy="0.0" ixz="0.0" iyy="100.11" iyz="0.0" izz="1.01" />
            <mass value="101.0" />
        </inertial>
        <visual>
            <origin xyz="0 0.0 0.01" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.1" length="0.02"/>
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <origin xyz="0 0.0 0.01" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.1" length="0.02"/>
            </geometry>
        </collision>
        <visual>
            <origin xyz="0 0.0 0.3" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.02" length="0.6"/>
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <origin xyz="0 0.0 0.3" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.02" length="0.6"/>
            </geometry>
        </collision>
    </link>


    <!-- link 1 -->
    <link name='link_1'>
        <inertial>
            <origin xyz="-0.00 -0.00 ${length/2}" rpy="0 0 0" />
            <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0104" iyz="0.0" izz="0.0" />
            <mass value="${mass}" />
        </inertial>
        <visual>
            <origin xyz="0 0 ${length/2}" rpy="0 0 0" />
            <geometry>
                <cylinder radius="${radius}" length="${length}"/>
            </geometry>
            <material name="orange"/>
        </visual>
        <collision>
            <origin xyz="0 0 ${length/2}" rpy="0 0 0" />
            <geometry>
                <cylinder radius="${radius}" length="${length}"/>
            </geometry>
        </collision>
    </link>

    <joint name="joint_1" type="continuous">
        <origin xyz="0 0 0.6" rpy= "0 -1.57 0" />
        <child link="link_1"/>
        <parent link="base_link" />
        <axis xyz="0 1 0" />
        <dynamics damping="${damping}" friction="0.0"/>
        <limit effort="${effort_limit}" velocity="${velocity_limit}"/>
    </joint>

    <!-- link 2 -->
    <link name='link_2'>
        <inertial>
            <origin xyz="-0.00 -0.00 0.125" rpy="0 0 0" />
            <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0104" iyz="0.0" izz="0.0" />
            <mass value="${mass}" />
        </inertial>
        <visual>
            <origin xyz="0 0 0.125" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.02" length="0.25"/>
            </geometry>
            <material name="red"/>
        </visual>
        <collision>
            <origin xyz="0 0 0.125" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.02" length="0.25"/>
            </geometry>
        </collision>
    </link>

    <joint name="joint_2" type="continuous">
        <origin xyz="0 0 ${length}" rpy= "0 0 0" />
        <child link="link_2"/>
        <parent link="link_1" />
        <axis xyz="0 1 0" />
        <dynamics damping="${damping}" />
        <limit effort="${effort_limit}" velocity="${velocity_limit}"/>
    </joint>



    <transmission name="tran1">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_1">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor1">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <transmission name="tran2">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_2">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor2">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    
</robot>